nupic/research/frameworks/pytorch/functions/mish.py

```diff
@@ -0,0 +1,34 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import torch
+import torch.nn.functional as F
+
+
+def mish(x):
+    """
+    Applies the mish function element-wise:
+    mish(x) = x * tanh(softplus(x)) = x * tanh(ln(1 + exp(x)))
+
+    Code adapted from:
+       https://github.com/digantamisra98/Mish/blob/master/Mish/Torch/functional.py
+    """
+    return x * torch.tanh(F.softplus(x))
```

Comment:
Is there an option to do in_place as relu is done, since it is not dependent on any variable? We found a significant increasing performance switching to inplace=True in relu, wondering if we will lose this switching to an inhouse implementation.

For reference, the C implementation of PyTorch activation functions is here: https://github.com/pytorch/pytorch/blob/34b32ca91491147cca2418b955141944d4492bec/aten/src/ATen/native/Activation.cpp

Reply:
This is a good point. Unfortunately, mish cannot be done in place. Inspecting the function `x * tanh(softplus(x))` shows you need at least two copies of the network. I may edit this in future iterations to use as little GPU memory as possible.